cloudflare/resource_cloudflare_teams_rules.go

```diff
@@ -0,0 +1,395 @@
+		Delete: resourceCloudflareTeamsRuleDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareTeamsRuleImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"account_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"precedence": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"action": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice(teamsRuleActions, false),
+				Required:     true,
+			},
+			"filters": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"traffic": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"identity": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"version": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"rule_settings": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: teamsRuleSettings,
+				},
+			},
+		},
+	}
+}
+
+var teamsRuleSettings = map[string]*schema.Schema{
+	"block_page_enabled": {
+		Type:     schema.TypeBool,
+		Optional: true,
+	},
+	"block_page_reason": {
+		Type:     schema.TypeString,
+		Optional: true,
+	},
+	"override_ips": {
+		Type:     schema.TypeList,
+		Optional: true,
+		Elem:     &schema.Schema{Type: schema.TypeString},
+	},
+	"override_host": {
+		Type:     schema.TypeString,
+		Optional: true,
+	},
+	"l4override": {
+		Type:     schema.TypeList,
+		MaxItems: 1,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: teamsL4OverrideSettings,
+		},
+	},
+	"biso_admin_controls": {
+		Type:     schema.TypeList,
+		MaxItems: 1,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: teamsBisoAdminControls,
+		},
+	},
+}
+
+var teamsL4OverrideSettings = map[string]*schema.Schema{
+	"ip": {
+		Type:     schema.TypeString,
+		Required: true,
+	},
+	"port": {
+		Type:     schema.TypeInt,
+		Required: true,
+	},
+}
+
+var teamsBisoAdminControls = map[string]*schema.Schema{
+	"disable_printing": {
+		Type:     schema.TypeBool,
+		Optional: true,
+	},
+	"disable_copy_paste": {
+		Type:     schema.TypeBool,
+		Optional: true,
+	},
+}
+
+func resourceCloudflareTeamsRuleRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	accountID := d.Get("account_id").(string)
+
+	rule, err := client.TeamsRule(context.Background(), accountID, d.Id())
+	if err != nil {
+		if strings.Contains(err.Error(), "HTTP status 400") {
+			log.Printf("[INFO] Teams Rule config %s doesnt exists", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error finding Teams Rule %q: %s", d.Id(), err)
+	}
+	if err := d.Set("name", rule.Name); err != nil {
```

Comment:
nice error handling on all these fields. basic field types (strings, numbers, booleans) don't really need it but it's good to have either way. the real win is the complex types and if we have malformed data going into it.